### PR TITLE
[RFC] Documentation Generators

### PR DESCRIPTION
**Pull request by KaemonIsland**
I'll update this comment once the docs are up on `gh-pages`. This is my first RFC, so please let me know if there is any additional information that I should add, or if you have any questions that were unanswered!

### CODE REVIEW

File: mkdocs.yml [hunk 1]
```diff
@@ -6,6 +6,7 @@ nav:
   - Deployment: deployment.md
   - Running Locally: running-locally.md
   - Backstage Update: backstage-update.md
+  - RFC Doc Generator: rfc-doc-generator.md
```
Comment (rianfowler) [inline mkdocs.yml:9]: I think we can add a folder for RFCs like this 
```suggestion
  RFCs and decision records: 
    - RFC Doc Generator: rfc-doc-generator.md
```

File: docs/rfc-doc-generator.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ * @return {string} This is the result
+ */
+function stringsStringStrings(p1, p2, p3, p4) {
+  // TODO
+}
+```
+
+Unfortunately, JSDoc doesn't support TypeScript out of the box. There are some plugins like [better-docs](https://github.com/SoftwareBrothers/better-docs) that can parse `.ts | .tsx` files. However, types aren't displayed correctly within the generated documentation. Meaning that we'd have to list out types within the comments and within the code. There is a way around it (I've done it before) but it requires some extra time and monkey-patching to get it to work correctly.
+
+## TypeDoc
```
Comment (rianfowler) [inline docs/rfc-doc-generator.md:59]: - What is the process for generating these documents? Is it just a CLI command that outputs html? 
- In what way do you think would work with [TechDocs](https://backstage.io/docs/features/techdocs/techdocs-overview) generation? 
- What kind of enforcement do you propose we start off with in eslint? Errors or warnings? 
- What kind of migration effort do you anticipate?

**Review by keyluck — APPROVED** [review]

File: docs/rfc-doc-generator.md (hunk 2 above)
Comment (KaemonIsland) [reply docs/rfc-doc-generator.md:71]: I've updated the docs to include this information.

tldr;

* You can use a CLI with options, or use a config file. `jsdoc.json` for JSDoc and `tsconfig.json` for TypeDoc.
* I don't think we can combine docs. So I think we should keep techdocs on GitHub Pages, then possibly use Netlify to host our docs or open a new repository that we publish to in a GitHub Action and use GitHub Pages there.
* I'm used to enforcing comments on all `Functions` (description, params, return) and everything else is optional. I'm not sure if we'd want to do a Warning for `Types` or `Interfaces`?
* Not sure what you mean by migration effort. Like how long it would take to add comments to all the current code in Backstage? For now, we could also use something like `lint-staged` to ensure new code follows the commenting rules? Then on our `Funday Fridays` we could work on commenting the existing code? Or add coments whenever we update a file that has uncommented code.

Comment (rianfowler) [reply docs/rfc-doc-generator.md:71]: > You can use a CLI with options, or use a config file. jsdoc.json for JSDoc and tsconfig.json for TypeDoc. 

sounds good 

>I don't think we can combine docs. So I think we should keep techdocs on GitHub Pages, then possibly use Netlify to host our docs or open a new repository that we publish to in a GitHub Action and use GitHub Pages there.

Eventually the tech docs will be published to our backstage instance. What is the output of this CLI command? 

TechDocs publication will push html files to an s3 bucket and the tech docs plugin makes those available in the backstage instance. Do you think we could take the documentation generated by this process and add it to the same s3 bucket as a test? 

>I'm used to enforcing comments on all Functions (description, params, return) and everything else is optional. I'm not sure if we'd want to do a Warning for Types or Interfaces?

I think enforcing comments on all functions is a good start and we can build on that. 

> Not sure what you mean by migration effort. Like how long it would take to add comments to all the current code in Backstage? For now, we could also use something like lint-staged to ensure new code follows the commenting rules? Then on our Funday Fridays we could work on commenting the existing code? Or add coments whenever we update a file that has uncommented code.

I wasn't sure if we would need to add a lot of comments to the existing backstage code. I think starting with the staged changes is a good idea and we could measure how many updates are needed if we wanted to enforce it on the whole application.

**Comment by rianfowler** [comment]
You can merge this in. we can keep the thread going on a closed PR.

File: docs/rfc-doc-generator.md (hunk 2 above)
Comment (KaemonIsland) [reply docs/rfc-doc-generator.md:71]: > TechDocs publication will push html files to an s3 bucket and the tech docs plugin makes those available in the backstage instance. Do you think we could take the documentation generated by this process and add it to the same s3 bucket as a test?

Ya know I already forgot about that. I think we would definitely be able to add the documentation to the s3 bucket.

Comment (rianfowler) [reply docs/rfc-doc-generator.md:71]: Sounds good- I think we'll need to test to see if it works. I don't know what kind of transformation the tech docs backend plugin does to the html files so it might come out weird. Thanks for looking into this.
